### ISSUE TITLE
Traffic Ops should ping the database on startup

### ISSUE THREAD

**Issue by rob05c** [apache/trafficcontrol#2649]
Every time I start TO and forget to start my DB dockerfile, it just starts up, everything appears to work, the login webpage works, until you actually try to log in, then you get a cryptic error.

I think TO itself (and probably TP) should do a database ping on startup, and refuse to start and log an informative error, if the DB can't be reached.

That will make it immediately obvious what's wrong, immediately obvious _that_ something is wrong; and also easier to orchestrate, automate, alarm, etc when the DB can't be reached initially.

**Comment by mitchell852** [comment]
I think TP should not be involved in determining if the TODB is running but this does make sense imo:

>TO refuse to start and log an informative error, if the DB can't be reached.

**Comment by rob05c** [comment]
Hm, I think I might've changed my mind on this.

It's really handy to have TO _not_ require the DB to start, and only when you actually make a request.

It makes debugging and developing much easier, not having to start the DB first and guarantee it's running.

But I think it also makes the TO service more resilient. It means you don't have a required startup order, TO itself can run and function independently of the database, and the DB can not exist on startup, go down and back up at any time, etc. And it's only an issue if the DB is unavailable when a request is actually made. And only affects that particular request, not the TO Service itself.

I think all that is valuable, and makes TO more robust and resilient.

But a warning/error log message on startup wouldn't be bad.

**Comment by rob05c** [comment]
Incidentally, the same applies to TP: Traffic Portal is more robust and resilient as a service, if it doesn't require TO, but can run and function as a service, and TO not existing only affects individual requests, not the service itself. (I believe all that's true today.)

**Comment by rawlinp** [comment]
I agree with Rob's most recent comments. It's better to log an error on startup if the DB can't be reached and allow TO to continue to run.

**Comment by mitchell852** [comment]
> Incidentally, the same applies to TP: Traffic Portal is more robust and resilient as a service, if it doesn't require TO, but can run and function as a service, and TO not existing only affects individual requests, not the service itself. (I believe all that's true today.)

yep, that's true re: TP. it can run w/o a TO but it won't be very useable...
